### PR TITLE
CXX-760 Don't rely on pkg-config and get the basic build working on windows

### PR DESCRIPTION
**Pull request by acmorrow**
Lots of CMake improvements needed to get the basic driver build process up and running on Windows. While doing so, made various required compile fixes as pointed out by MSVC. Also adds support for building and installing CMake compatible find_package support for bsoncxx and mongocxx.

This plus the boost changes in https://github.com/mongodb/mongo-cxx-driver/pull/410 get you a more or less working build environment on Windows. There are still many problems there. In particular VS2015's compiler is ICE-ier than a skating rink and rejects a lot of valid code. The great renaming will fix some of those, but not others, and not the ICEs.

I'm not really expecting anyone to dig into this too closely unless you are really fascinated by CMake. I'll probably push that and the boost changes TBR since this is a long weekend.

### CODE REVIEW

File: cmake/FindLibBSON.cmake [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# Load up PkgConfig if we have it
+find_package(PkgConfig QUIET)
+
+if (PKG_CONFIG_FOUND)
+  pkg_check_modules(LIBBSON REQUIRED libbson-1.0>=${LibBSON_FIND_VERSION} )
+  # We don't reiterate the version information here because we assume that
+  # pkg_check_modules has honored our request.
+  find_package_handle_standard_args(LIBBSON DEFAULT_MSG LIBBSON_FOUND)
+elseif(LIBBSON_DIR)
+  # The best we can do until libbson starts instaling a libbson-config.cmake file
```
Comment (samantharitter) [inline cmake/FindLibBSON.cmake]: typo, "installing"

File: cmake/FindLibBSON.cmake [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  set(LIBBSON_LIBRARIES bson-1.0 CACHE INTERNAL "")
+  set(LIBBSON_LIBRARY_DIRS ${LIBBSON_DIR}/lib CACHE INTERNAL "")
+  set(LIBBSON_INCLUDE_DIRS ${LIBBSON_DIR}/include/libbson-1.0 CACHE INTERNAL "")
+  find_package_handle_standard_args(LIBBSON DEFAULT_MSG LIBBSON_LIBRARIES LIBBSON_LIBRARY_DIRS LIBBSON_INCLUDE_DIRS)
+else()
+  message(FATAL_ERROR "Don't know how to find libbson, please set LIBBSON_DIR")
```
Comment (markbenvenuto) [inline cmake/FindLibBSON.cmake]: what should it be set to?

File: src/bsoncxx/CMakeLists.txt [hunk 3]
```diff
@@ -64,8 +64,17 @@ add_library(bsoncxx_static STATIC
 
 set_target_properties(bsoncxx_static PROPERTIES
     OUTPUT_NAME bsoncxx
+    COMPILE_DEFINITIONS "BSONCXX_STATIC"
```
Comment (markbenvenuto) [inline src/bsoncxx/CMakeLists.txt]: Use target_compile_definitions(bsoncxx_static PUBLIC BSONCXX_STATIC), it is clearer

File: src/bsoncxx/config/compiler.hpp [hunk 4]
```diff
@@ -12,4 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#if defined(_MSC_VER)
+
+#pragma warning(push)
+#pragma warning(disable: 4251 4273 4275)
```
Comment (markbenvenuto) [inline src/bsoncxx/config/compiler.hpp]: What do these warnings mean?

File: cmake/FindLibMongoC.cmake [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+# Load up PkgConfig if we have it
+find_package(PkgConfig QUIET)
+
+if (PKG_CONFIG_FOUND)
+  pkg_check_modules(LIBMONGOC REQUIRED libmongoc-1.0>=${LibMongoC_FIND_VERSION} )
+  # We don't reiterate the version information here because we assume that
+  # pkg_check_modules has honored our request.
+  find_package_handle_standard_args(LIBMONGOC DEFAULT_MSG LIBMONGOC_FOUND)
+elseif(LIBMONGOC_DIR)
+  # The best we can do until libMONGOC starts instaling a libmongoc-config.cmake file
```
Comment (samantharitter) [inline cmake/FindLibMongoC.cmake]: ditto typo

File: src/bsoncxx/config/compiler.hpp (hunk 4 above)
Comment (samantharitter) [inline src/bsoncxx/config/compiler.hpp]: Can you add a comment explaining which warnings we are suppressing and why?

File: src/bsoncxx/private/error_category.cpp [hunk 6]
```diff
@@ -137,15 +137,18 @@ const std::error_category& bsoncxx_error_category() {
 
 const std::error_category& libbson_error_category(int domain) {
     switch (domain) {
-        case BSON_ERROR_JSON:
+        case BSON_ERROR_JSON: {
```
Comment (samantharitter) [inline src/bsoncxx/private/error_category.cpp]: Why do we need { } here?

File: src/bsoncxx/config/compiler.hpp (hunk 4 above)
Comment (samantharitter) [inline src/bsoncxx/config/compiler.hpp]: Oh I must not have refreshed since you added your comment, Mark

File: cmake/FindLibBSON.cmake (hunk 1 above)
Comment (acmorrow) [inline cmake/FindLibBSON.cmake]: Done

File: cmake/FindLibBSON.cmake (hunk 2 above)
Comment (acmorrow) [inline cmake/FindLibBSON.cmake]: Done

File: cmake/FindLibMongoC.cmake (hunk 5 above)
Comment (acmorrow) [inline cmake/FindLibMongoC.cmake]: Done

File: cmake/FindLibMongoC.cmake [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  set(LIBMONGOC_LIBRARIES mongoc-1.0 CACHE INTERNAL "")
+  set(LIBMONGOC_LIBRARY_DIRS ${LIBMONGOC_DIR}/lib CACHE INTERNAL "")
+  set(LIBMONGOC_INCLUDE_DIRS ${LIBMONGOC_DIR}/include/libMONGOC-1.0 CACHE INTERNAL "")
+  find_package_handle_standard_args(LIBMONGOC DEFAULT_MSG LIBMONGOC_LIBRARIES LIBMONGOC_LIBRARY_DIRS LIBMONGOC_INCLUDE_DIRS)
+else()
+    message(FATAL_ERROR "Don't know how to find libmongoc, please set LIBMONGOC_DIR")
```
Comment (acmorrow) [inline cmake/FindLibMongoC.cmake]: Expanded message here too.

File: src/bsoncxx/private/error_category.cpp (hunk 6 above)
Comment (acmorrow) [inline src/bsoncxx/private/error_category.cpp]: This was actually a real bug that MSVC told us about. Without braces in each case, the scope of the static category extends to the entire switch statement, but you aren't allowed to jump past the initialization of static.

File: src/bsoncxx/CMakeLists.txt (hunk 3 above)
Comment (acmorrow) [inline src/bsoncxx/CMakeLists.txt]: Good call, done, here and for mongocxx

File: src/bsoncxx/config/compiler.hpp (hunk 4 above)
Comment (acmorrow) [inline src/bsoncxx/config/compiler.hpp]: Thanks for making me take another look at this. I think the only one we really need is 4251, and I've added a comment about that here and for mongocxx. I think the other ones were spurious due to misconfigurations addressed elsewhere in this PR.

**Comment by samantharitter** [comment]
LGTM with a pretty shallow swim into the cmake

**Comment by acmorrow** [comment]
Updated, PTAL. All review changes are in the newest commit.

**Comment by markbenvenuto** [comment]
LGTM

**Pull request merged**